File: yep/src/main/java/catchla/yep/fragment/TopicsListFragment.java
package catchla.yep.fragment;

import android.accounts.Account;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Rect;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;

import java.util.List;

import catchla.yep.R;
import catchla.yep.activity.LocationPickerActivity;
import catchla.yep.activity.MediaViewerActivity;
import catchla.yep.activity.NewTopicActivity;
import catchla.yep.activity.SkillUpdatesActivity;
import catchla.yep.activity.TopicChatActivity;
import catchla.yep.activity.UserActivity;
import catchla.yep.adapter.TopicsAdapter;
import catchla.yep.fragment.iface.IActionButtonSupportFragment;
import catchla.yep.loader.DiscoverTopicsLoader;
import catchla.yep.model.Attachment;
import catchla.yep.model.Paging;
import catchla.yep.model.Topic;
import catchla.yep.view.holder.TopicViewHolder;

/**
 * Created by mariotaku on 15/10/12.
 */
public class TopicsListFragment extends AbsContentListRecyclerViewFragment<TopicsAdapter>
        implements LoaderManager.LoaderCallbacks<List<Topic>>, TopicsAdapter.TopicClickListener,
        IActionButtonSupportFragment {

    private String mSortBy;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mSortBy = mPreferences.getString(KEY_TOPICS_SORT_ORDER, Topic.SortOrder.TIME);
        setHasOptionsMenu(true);
        final Bundle fragmentArgs = getArguments();
        final Bundle loaderArgs = new Bundle();
        if (fragmentArgs != null) {
            loaderArgs.putBoolean(EXTRA_READ_CACHE, !fragmentArgs.containsKey(EXTRA_LEARNING)
                    && !fragmentArgs.containsKey(EXTRA_MASTER));
            if (fragmentArgs.containsKey(EXTRA_USER_ID)) {
                loaderArgs.putString(EXTRA_USER_ID, fragmentArgs.getString(EXTRA_USER_ID));
            }
        } else {
            loaderArgs.putBoolean(EXTRA_READ_CACHE, true);
        }
        getLoaderManager().initLoader(0, loaderArgs, this);
        getAdapter().setClickListener(this);
        showProgress();
    }

    @Override
    public Loader<List<Topic>> onCreateLoader(final int id, final Bundle args) {
        final boolean cachingEnabled = isCachingEnabled();
        final boolean readCache = args.getBoolean(EXTRA_READ_CACHE) && cachingEnabled;
        final boolean readOld = args.getBoolean(EXTRA_READ_OLD, readCache) && cachingEnabled;
        final String maxId = args.getString(EXTRA_MAX_ID);
        final Paging paging = new Paging();
        if (maxId != null) {
            paging.maxId(maxId);
        }
        final List<Topic> oldData;
        if (readOld) {
            oldData = getAdapter().getTopics();
        } else {
            oldData = null;
        }
        return new DiscoverTopicsLoader(getActivity(), getAccount(), getArguments().getString(EXTRA_USER_ID),
                paging, getSortOrder(), readCache, cachingEnabled, oldData);
    }

    @Topic.SortOrder
    private String getSortOrder() {
        if (hasUserId()) return Topic.SortOrder.TIME;
        return mSortBy != null ? mSortBy : Topic.SortOrder.TIME;
    }

    private boolean hasUserId() {
        final Bundle fragmentArgs = getArguments();
        return fragmentArgs != null && fragmentArgs.containsKey(EXTRA_USER_ID);
    }

    @Override
    public void onLoadFinished(final Loader<List<Topic>> loader, final List<Topic> data) {
        final TopicsAdapter adapter = getAdapter();
        adapter.setData(data);
        adapter.setLoadMoreSupported(data != null && !data.isEmpty());
        showContent();
        setRefreshing(false);
        setRefreshEnabled(true);
        setLoadMoreIndicatorVisible(false);
    }

    @Override
    public void onLoaderReset(final Loader<List<Topic>> loader) {

    }

    @NonNull
    @Override
    protected TopicsAdapter onCreateAdapter(Context context) {
        return new TopicsAdapter(context);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_fragment_chats_list, menu);
    }

    @Override
    public void onBaseViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onBaseViewCreated(view, savedInstanceState);
    }

    @Override
    public void onRefresh() {
        final Bundle loaderArgs = new Bundle();
        loaderArgs.putBoolean(EXTRA_READ_CACHE, false);
        getLoaderManager().restartLoader(0, loaderArgs, this);
    }

    @Override
    public boolean isRefreshing() {
        return getLoaderManager().hasRunningLoaders();
    }

    public boolean isCachingEnabled() {
        return getArguments().getBoolean(EXTRA_CACHING_ENABLED);
    }

    @Override
    public void onItemClick(final int position, final RecyclerView.ViewHolder holder) {
        final Topic topic = getAdapter().getTopic(position);
        final Intent intent = new Intent(getActivity(), TopicChatActivity.class);
        intent.putExtra(EXTRA_ACCOUNT, getAccount());
        intent.putExtra(EXTRA_TOPIC, topic);
        startActivity(intent);
    }

    private Account getAccount() {
        return getArguments().getParcelable(EXTRA_ACCOUNT);
    }

    @Override
    public int getActionIcon() {
        return R.drawable.ic_action_edit;
    }

    @Override
    public void onActionPerformed() {
        final Bundle args = new Bundle();
        args.putParcelable(EXTRA_ACCOUNT, getAccount());
        final NewTopicTypeDialogFragment df = new NewTopicTypeDialogFragment();
        df.setArguments(args);
        df.show(getFragmentManager(), "new_topic_type");
    }

    @Nullable
    @Override
    public Class<? extends FloatingActionMenuFragment> getActionMenuFragment() {
        return TopicsMenuFragment.class;
    }

    @Override
    public void onLoadMoreContents() {
        super.onLoadMoreContents();
        final Bundle loaderArgs = new Bundle();
        loaderArgs.putBoolean(EXTRA_READ_CACHE, false);
        loaderArgs.putBoolean(EXTRA_READ_OLD, true);
        final TopicsAdapter adapter = getAdapter();
        final int topicsCount = adapter.getTopicsCount();
        if (topicsCount > 0) {
            loaderArgs.putString(EXTRA_MAX_ID, adapter.getTopic(topicsCount - 1).getId());
        }
        getLoaderManager().restartLoader(0, loaderArgs, this);
    }

    @Override
    public void onSkillClick(final int position, final TopicViewHolder holder) {
        final Intent intent = new Intent(getContext(), SkillUpdatesActivity.class);
        intent.putExtra(EXTRA_ACCOUNT, getAccount());
        intent.putExtra(EXTRA_SKILL, getAdapter().getTopic(position).getSkill());
        startActivity(intent);
    }

    @Override
    public void onUserClick(final int position, final TopicViewHolder holder) {
        final Intent intent = new Intent(getContext(), UserActivity.class);
        intent.putExtra(EXTRA_ACCOUNT, getAccount());
        intent.putExtra(EXTRA_USER, getAdapter().getTopic(position).getUser());
        startActivity(intent);
    }

    @Override
    public void onMediaClick(final Attachment[] attachments, final Attachment attachment, View view) {
        final Intent intent = new Intent(getContext(), MediaViewerActivity.class);
        intent.putExtra(EXTRA_MEDIA, attachments);
        intent.putExtra(EXTRA_CURRENT_MEDIA, attachment);
        int[] location = new int[2];
        view.getLocationOnScreen(location);
        intent.setSourceBounds(new Rect(location[0], location[1], location[0] + view.getWidth(),
                location[1] + view.getHeight()));
        final Bundle options = ActivityOptionsCompat.makeScaleUpAnimation(view, 0, 0,
                view.getWidth(), view.getHeight()).toBundle();
        ActivityCompat.startActivity(getActivity(), intent, options);
    }

    public void reloadWithSortOrder(final String sortBy) {
        if (TextUtils.equals(getSortOrder(), sortBy) || hasUserId()) return;
        mSortBy = sortBy;
        mPreferences.edit().putString(KEY_TOPICS_SORT_ORDER, sortBy).apply();
        final Bundle loaderArgs = new Bundle();
        loaderArgs.putBoolean(EXTRA_READ_CACHE, false);
        loaderArgs.putBoolean(EXTRA_READ_OLD, false);
        getLoaderManager().restartLoader(0, loaderArgs, this);
        showProgress();
    }

    public static class NewTopicTypeDialogFragment extends DialogFragment {
        @NonNull
        @Override
        public Dialog onCreateDialog(final Bundle savedInstanceState) {
            final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            final Resources resources = getResources();
            final String[] entries = resources.getStringArray(R.array.new_topic_type_entries);
            final String[] values = resources.getStringArray(R.array.new_topic_type_values);
            builder.setItems(entries, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(final DialogInterface dialog, final int which) {
                    switch (values[which]) {
                        case "photos_text": {
                            final Intent intent = new Intent(getContext(), NewTopicActivity.class);
                            intent.putExtra(EXTRA_ACCOUNT, getAccount());
                            startActivity(intent);
                            break;
                        }
                        case "audio": {
                            break;
                        }
                        case "location": {
                            final Intent intent = new Intent(getContext(), LocationPickerActivity.class);
                            intent.putExtra(EXTRA_ACCOUNT, getAccount());
                            startActivity(intent);
                            break;
                        }
                    }
                }
            });
            return builder.create();
        }

        private Account getAccount() {
            return getArguments().getParcelable(EXTRA_ACCOUNT);
        }
    }
}

File: src/main/java/catchla/yep/fragment/TopicsListFragment.kt
package catchla.yep.fragment

import android.accounts.Account
import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Rect
import android.location.Location
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.app.DialogFragment
import android.support.v4.app.LoaderManager
import android.support.v4.content.Loader
import android.support.v7.app.AlertDialog
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import catchla.yep.Constants
import catchla.yep.R
import catchla.yep.activity.*
import catchla.yep.adapter.TopicsAdapter
import catchla.yep.adapter.iface.ILoadMoreSupportAdapter.IndicatorPosition
import catchla.yep.fragment.iface.IActionButtonSupportFragment
import catchla.yep.loader.DiscoverTopicsLoader
import catchla.yep.model.Attachment
import catchla.yep.model.LocationAttachment
import catchla.yep.model.Paging
import catchla.yep.model.Topic
import catchla.yep.view.holder.TopicViewHolder

/**
 * Created by mariotaku on 15/10/12.
 */
class TopicsListFragment : AbsContentListRecyclerViewFragment<TopicsAdapter>(), LoaderManager.LoaderCallbacks<List<Topic>>, TopicsAdapter.TopicClickListener, IActionButtonSupportFragment {

    @Topic.SortOrder
    private var mSortBy: String? = null

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        //noinspection WrongConstant
        mSortBy = mPreferences.getString(Constants.KEY_TOPICS_SORT_ORDER, Topic.SortOrder.DEFAULT)
        setHasOptionsMenu(true)
        val fragmentArgs = arguments
        val loaderArgs = Bundle()
        if (fragmentArgs != null) {
            loaderArgs.putBoolean(Constants.EXTRA_READ_CACHE, !fragmentArgs.containsKey(Constants.EXTRA_LEARNING) && !fragmentArgs.containsKey(Constants.EXTRA_MASTER))
            if (fragmentArgs.containsKey(Constants.EXTRA_USER_ID)) {
                loaderArgs.putString(Constants.EXTRA_USER_ID, fragmentArgs.getString(Constants.EXTRA_USER_ID))
            }
        } else {
            loaderArgs.putBoolean(Constants.EXTRA_READ_CACHE, true)
        }
        loaderManager.initLoader(0, loaderArgs, this)
        adapter.clickListener = this
        showProgress()
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<List<Topic>> {
        val cachingEnabled = isCachingEnabled
        val readCache = args!!.getBoolean(Constants.EXTRA_READ_CACHE) && cachingEnabled
        val readOld = args.getBoolean(Constants.EXTRA_READ_OLD, readCache) && cachingEnabled
        val maxId = args.getString(Constants.EXTRA_MAX_ID)
        val paging = Paging()
        if (maxId != null) {
            paging.maxId(maxId)
        }
        val oldData: List<Topic>?
        if (readOld) {
            oldData = adapter.topics
        } else {
            oldData = null
        }
        return DiscoverTopicsLoader(activity, account, arguments.getString(Constants.EXTRA_USER_ID),
                paging, sortOrder, readCache, cachingEnabled, oldData)
    }

    private val sortOrder: String
        @Topic.SortOrder
        get() {
            if (hasUserId()) return Topic.SortOrder.TIME
            return if (mSortBy != null) mSortBy!! else Topic.SortOrder.TIME
        }

    private fun hasUserId(): Boolean {
        val fragmentArgs = arguments
        return fragmentArgs != null && fragmentArgs.containsKey(Constants.EXTRA_USER_ID)
    }

    override fun onLoadFinished(loader: Loader<List<Topic>>, data: List<Topic>?) {
        val adapter = adapter
        adapter.topics = data
        adapter.loadMoreSupportedPosition = if (data != null && !data.isEmpty()) IndicatorPosition.END else IndicatorPosition.NONE
        showContent()
        isRefreshing = false
        setRefreshEnabled(true)
        setLoadMoreIndicatorPosition(IndicatorPosition.NONE)
    }

    override fun onLoaderReset(loader: Loader<List<Topic>>) {

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_NEW_LOCATION_TOPIC -> {
                if (resultCode == Activity.RESULT_OK) {
                    val name = data!!.getStringExtra(Constants.EXTRA_NAME)
                    val location = data.getParcelableExtra<Location>(Constants.EXTRA_LOCATION)
                    val intent = Intent(context, NewTopicActivity::class.java)
                    val attachment = LocationAttachment()
                    attachment.place = name
                    attachment.latitude = location.latitude
                    attachment.longitude = location.longitude
                    intent.putExtra(Constants.EXTRA_NEW_TOPIC_TYPE, NewTopicActivity.TYPE_LOCATION)
                    intent.putExtra(Constants.EXTRA_ATTACHMENT, attachment)
                    intent.putExtra(Constants.EXTRA_ACCOUNT, account)
                    startActivity(intent)
                    return
                }
                return
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onCreateAdapter(context: Context): TopicsAdapter {
        return TopicsAdapter(context)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.menu_fragment_chats_list, menu)
    }

    override fun onBaseViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onBaseViewCreated(view, savedInstanceState)
    }

    override fun onRefresh() {
        val loaderArgs = Bundle()
        loaderArgs.putBoolean(Constants.EXTRA_READ_CACHE, false)
        loaderManager.restartLoader(0, loaderArgs, this)
    }

    override fun isRefreshing(): Boolean {
        return loaderManager.hasRunningLoaders()
    }

    val isCachingEnabled: Boolean
        get() = arguments.getBoolean(Constants.EXTRA_CACHING_ENABLED)

    override fun onItemClick(position: Int, holder: RecyclerView.ViewHolder) {
        val topic = adapter.getTopic(position)
        val intent = Intent(activity, TopicChatActivity::class.java)
        intent.putExtra(Constants.EXTRA_ACCOUNT, account)
        intent.putExtra(Constants.EXTRA_TOPIC, topic)
        startActivity(intent)
    }

    private val account: Account
        get() = arguments.getParcelable<Account>(Constants.EXTRA_ACCOUNT)

    override fun getActionIcon(): Int {
        return R.drawable.ic_action_edit
    }

    override fun onActionPerformed() {
        val args = Bundle()
        args.putParcelable(Constants.EXTRA_ACCOUNT, account)
        val df = NewTopicTypeDialogFragment()
        df.arguments = args
        df.show(childFragmentManager, "new_topic_type")
    }

    override fun getActionMenuFragment(): Class<out FloatingActionMenuFragment>? {
        return TopicsMenuFragment::class.java
    }

    override fun onLoadMoreContents(@IndicatorPosition position: Int) {
        // Only supports load from end, skip START flag
        if (position and IndicatorPosition.START != 0) return
        super.onLoadMoreContents(position)
        val loaderArgs = Bundle()
        loaderArgs.putBoolean(Constants.EXTRA_READ_CACHE, false)
        loaderArgs.putBoolean(Constants.EXTRA_READ_OLD, true)
        val adapter = adapter
        val topicsCount = adapter.topicsCount
        if (topicsCount > 0) {
            loaderArgs.putString(Constants.EXTRA_MAX_ID, adapter.getTopic(topicsCount - 1).id)
        }
        loaderManager.restartLoader(0, loaderArgs, this)
    }

    override fun onSkillClick(position: Int, holder: TopicViewHolder) {
        val intent = Intent(context, SkillUpdatesActivity::class.java)
        intent.putExtra(Constants.EXTRA_ACCOUNT, account)
        intent.putExtra(Constants.EXTRA_SKILL, adapter.getTopic(position).skill)
        startActivity(intent)
    }

    override fun onUserClick(position: Int, holder: TopicViewHolder) {
        val intent = Intent(context, UserActivity::class.java)
        intent.putExtra(Constants.EXTRA_ACCOUNT, account)
        intent.putExtra(Constants.EXTRA_USER, adapter.getTopic(position).user)
        startActivity(intent)
    }

    override fun onMediaClick(attachments: Array<Attachment>, attachment: Attachment, clickedView: View) {
        val intent = Intent(context, MediaViewerActivity::class.java)
        intent.putExtra(Constants.EXTRA_MEDIA, attachments)
        intent.putExtra(Constants.EXTRA_CURRENT_MEDIA, attachment)
        val location = IntArray(2)
        clickedView.getLocationOnScreen(location)
        intent.sourceBounds = Rect(location[0], location[1], location[0] + clickedView.width,
                location[1] + clickedView.height)
        val options = ActivityOptionsCompat.makeScaleUpAnimation(clickedView, 0, 0,
                clickedView.width, clickedView.height).toBundle()
        ActivityCompat.startActivity(activity, intent, options)
    }

    fun reloadWithSortOrder(@Topic.SortOrder sortBy: String) {
        if (TextUtils.equals(sortOrder, sortBy) || hasUserId()) return
        mSortBy = sortBy
        mPreferences.edit().putString(Constants.KEY_TOPICS_SORT_ORDER, sortBy).apply()
        val loaderArgs = Bundle()
        loaderArgs.putBoolean(Constants.EXTRA_READ_CACHE, false)
        loaderArgs.putBoolean(Constants.EXTRA_READ_OLD, false)
        loaderManager.restartLoader(0, loaderArgs, this)
        showProgress()
    }

    class NewTopicTypeDialogFragment : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val builder = AlertDialog.Builder(context)
            val resources = resources
            val entries = resources.getStringArray(R.array.new_topic_type_entries)
            val values = resources.getStringArray(R.array.new_topic_type_values)
            builder.setItems(entries) { dialog, which ->
                when (values[which]) {
                    "photos_text" -> {
                        val intent = Intent(context, NewTopicActivity::class.java)
                        intent.putExtra(Constants.EXTRA_ACCOUNT, account)
                        startActivity(intent)
                    }
                    "audio" -> {
                    }
                    "location" -> {
                        val parent = parentFragment
                        val intent = Intent(context, LocationPickerActivity::class.java)
                        intent.putExtra(Constants.EXTRA_ACCOUNT, account)
                        parent.startActivityForResult(intent, REQUEST_NEW_LOCATION_TOPIC)
                    }
                }
            }
            return builder.create()
        }

        private val account: Account
            get() = arguments.getParcelable<Account>(Constants.EXTRA_ACCOUNT)
    }

    companion object {

        private val REQUEST_NEW_LOCATION_TOPIC = 102
    }
}
